### PR TITLE
Custom layout renderer to support nlog target async for HttpContext

### PR DESCRIPTION
**Pull request by kggayo**
-Added a custom layout renderer to capture HttpContext to support nlog target async
-Changed the HttpContext to HttpContextBase to support fullframework and .net core
-Usage: <contextproperty name="stackifyhttp" layout="${stackify-http}" />

Sample:
 <nlog>
    <extensions>
      <add assembly="NLog.Targets.Stackify" />
    </extensions>
    <targets async="true">
      <target type="StackifyTarget" name="stackify" logAllParams="true" >
        <contextproperty name="stackifyhttp" layout="${stackify-http}" />
      </target>
    </targets>
    <rules>
      <logger name="*" minlevel="Debug" writeTo="stackify" />
    </rules>
  </nlog>

### CODE REVIEW

**Comment by snakefoot** [comment]
Instead of abusing `contextproperty` with magic-string-values, then it would be better if you created a dedicated Layout-property on the NLog StackifyTarget:

Ex.:

```c#
class StackifyTarget 
{
     public Layout StackifyHttpRequest { get; set; }
}
```

This will also make it easier to include in the documentation.

You can optimize the `StackifyHttpLayoutRenderer` by adding the attribute. `[ThreadSafe]` (NLog 4.5.6 or newer):

```c#
   [LayoutRenderer("stackify-http")]
   [ThreadSafe]
   public class StackifyHttpLayoutRenderer : AspNetLayoutRendererBase
```

It will allow the NLog engine to perform the async-logging much more concurrently.

File: Src/NLog.Targets.Stackify/NLog.Targets.Stackify.csproj [hunk 1]
```diff
@@ -22,10 +22,12 @@
     <RepositoryUrl>https://github.com/stackify/stackify-api-dotnet</RepositoryUrl>
     <RepositoryType>git</RepositoryType>
   </PropertyGroup>
-
+  
   <ItemGroup>
     <ProjectReference Include="..\StackifyLib\StackifyLib.csproj" />
     <PackageReference Include="NLog" Version="4.5.0" />
+    <PackageReference Include="NLog.Config" Version="4.5.0" />
```
Comment (snakefoot) [inline Src/NLog.Targets.Stackify/NLog.Targets.Stackify.csproj]: Please remove `NLog.Config` as dependency.

File: Src/NLog.Targets.Stackify/NLog.Targets.Stackify.csproj [hunk 2]
```diff
@@ -35,5 +37,17 @@
...
+    <Reference Include="System.Web.Abstractions" />
+    <Reference Include="System.Web.Routing" />
+    <PackageReference Include="NLog.Web" Version="4.5.0" />
+  </ItemGroup>
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0'">
+    <PackageReference Include="Microsoft.AspNetCore.Http" Version="2.1.1" />
+    <PackageReference Include="Microsoft.AspNetCore.Hosting.Abstractions" Version="2.1.1" />
+    <PackageReference Include="Microsoft.AspNetCore.Routing.Abstractions" Version="2.1.1" />
+    <PackageReference Include="NLog.Web.AspNetCore" Version="4.5.0" />
+  </ItemGroup>
```
Comment (snakefoot) [inline Src/NLog.Targets.Stackify/NLog.Targets.Stackify.csproj]: Why have all these System.Web and AspNetCore dependencies been added to `NLog.Targets.Stackify` ? Please remove them again.

File: Src/NLog.Targets.Stackify/NLog.Targets.Stackify.csproj [hunk 3]
```diff
@@ -35,5 +37,17 @@
...
+    <Reference Include="System.Web" />
+    <Reference Include="System.Web.Abstractions" />
+    <Reference Include="System.Web.Routing" />
+    <PackageReference Include="NLog.Web" Version="4.5.0" />
+  </ItemGroup>
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0'">
+    <PackageReference Include="Microsoft.AspNetCore.Http" Version="2.1.1" />
+    <PackageReference Include="Microsoft.AspNetCore.Hosting.Abstractions" Version="2.1.1" />
+    <PackageReference Include="Microsoft.AspNetCore.Routing.Abstractions" Version="2.1.1" />
+    <PackageReference Include="NLog.Web.AspNetCore" Version="4.5.0" />
```
Comment (snakefoot) [inline Src/NLog.Targets.Stackify/NLog.Targets.Stackify.csproj]: If you need to depend on NLog.Web and NLog.Web.AspNetCore. Then you should choose version `4.8.0` (As it matches `NLog 4.5.0`)

**Comment by snakefoot** [comment]
Curious if Stackify is only limited to Web-Applications? This change forces all NetCore-applications to take dependency on AspNetCore.

**Comment by snakefoot** [comment]
@sporty81 Curious if Stackify is only limited to Web-Applications? This change forces all NetCore-applications to take dependency on AspNetCore (Ex. AWS Lambda functions writing to Stackify)

**Comment by kggayo** [comment]
@snakefoot thank you for the review. I will apply those comments above. 
As for the AspNetCore dependency, i will create a new project for this custom layout renderer.

**Comment by kggayo** [comment]
@snakefoot when i use **[ThreadSafe]** the custom layout never gets called. Are there any config that needs to be added to make it work? Thanks

**Comment by snakefoot** [comment]
Can you give a link to a repository with the non working code? I'm guessing that you fail to add your assembly to extensions and update the application to use NLog 4.5.3 or newer


Sent from my Sony Xperia


---- Kevin Gayo wrote ----


@snakefoot<https://github.com/snakefoot> when i use [ThreadSafe] the custom layout never gets called. Are there any config that needs to be added to make it work? Thanks

**Comment by kggayo** [comment]
@snakefoot i failed to update the app to use NLog 4.5.3. [ThreadSafe] now works with custom layout renderer. 
Thanks a lot for your help!

**Pull request merged**